intermediate_source/custom_function_conv_bn_tutorial.py

```diff
@@ -0,0 +1,344 @@
+# backward function, but below we first implement them as their own
+# custom functions so we can validate their correctness individually
+import torch
+from torch.autograd.function import once_differentiable
+import torch.nn.functional as F
+
+class Conv2D(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, X, weight):
+        ctx.save_for_backward(X, weight)
+        return F.conv2d(X, weight)
+
+    # Use @once_differentiable by default unless we intend to double
+    # backward
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_out):
+        X, weight = ctx.saved_tensors
+        grad_input = F.conv2d(X.transpose(0, 1), grad_out.transpose(0, 1)).transpose(0, 1)
+        grad_X = F.conv_transpose2d(grad_out, weight)
+        return grad_X, grad_input
+
+######################################################################
+# When testing with gradcheck, it is important to use double precision
+weight = torch.rand(5, 3, 3, 3, requires_grad=True, dtype=torch.double)
+X = torch.rand(10, 3, 7, 7, requires_grad=True, dtype=torch.double)
+torch.autograd.gradcheck(Conv2D.apply, (X, weight))
+
+######################################################################
+# Batch Norm
+def expand(t):
+    # Helper function to unsqueeze the dimensions that we reduce over
+    return t[None, :, None, None]
+
+def batch_norm_backward(grad_out, X, sum, sqrt_var, N, eps):
+    # To simplify our derivation, we follow the chain rule and compute
+    # the following gradients before accumulating them all into a final grad_input
+    #  1) X of the numerator
+    #  2) mean of the numerator
+    #  3) var of the denominator
+    # d_denom = -num / denom**2
+    tmp = -((X - expand(sum) / N) * grad_out).sum(dim=(0, 2, 3))
+    d_denom = tmp / (sqrt_var + eps)**2
+    # It is useful to delete tensors when you no longer need them
+    # It's not a big difference here though because tmp only has size of (C,)
+    # The important thing is avoid allocating NCHW-sized tensors unnecessarily
+    del tmp
+    grad_input = grad_out / expand(sqrt_var + eps)  # = d_numerator
+    grad_input -= expand(grad_input.sum(dim=(0, 2, 3))) / N  # mean = X.sum(dim=(0, 2, 3)) / N
+    # denom = torch.sqrt(var) + eps
+    d_var = d_denom / (2 * sqrt_var)
+    del d_denom
+    # unbiased_var(x) = ((X - expand(mean))**2).sum(dim=(0, 2, 3)) / (N - 1)
+    grad_input += (X * expand(d_var) * N + expand(-d_var * sum)) * 2 / ((N - 1) * N)
+    return grad_input
+
+class BatchNorm(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, X, eps=1e-6):
+        # Don't save keepdim'd values for backward
+        sum = X.sum(dim=(0, 2, 3))
+        var = X.var(unbiased=True, dim=(0, 2, 3))
+        N = X.numel() / X.size(1)
+        sqrt_var = torch.sqrt(var)
+        ctx.save_for_backward(X)
+        ctx.eps = eps
+        ctx.sum = sum
+        ctx.N = N
+        ctx.sqrt_var = sqrt_var
+        mean = sum / N
```

Comment:
Here as well, you can show that you can do everything inplace to reduce memory usage

Reply:
Not sure I can do this one in-place because it would modify ctx.sum as well, which I wanted to save for backward

Reply:
Yes sorry this comment is not pointing at the right line in the code.
The return statement below does contain some opportunities though